header-only/CMakeLists.txt

```diff
@@ -0,0 +1,71 @@
+# BSD 3-Clause License; see https://github.com/scikit-hep/awkward-1.0/blob/main/LICENSE
+
+cmake_minimum_required(VERSION 3.15...3.24)
+
+project(
+  awkward-headers
+  LANGUAGES CXX
+  VERSION 1.0.0)
+
+include(GNUInstallDirs)
+
+# Build tests?
+option(BUILD_TESTS OFF)
+
+# Add aliases for `add_subdirectory`
+set(NAMESPACE "awkward::")
+
+# We want to specify different compile options, so we split the headers into "components"
+macro(add_component name)
+  add_library(${name} INTERFACE)
+  add_library(${NAMESPACE}${name} ALIAS ${name})
+  target_include_directories(
+    ${name} INTERFACE $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/${name}>
+                      $<INSTALL_INTERFACE:include/${name}>)
+
+  install(DIRECTORY "${CMAKE_CURRENT_SOURCE_DIR}/${name}/awkward"
+          DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}/${name}")
+endmacro(add_component)
+
+# Component: Builder Options
+add_component(builder-options)
+# Component: Growable Buffer
+add_component(growable-buffer)
+target_link_libraries(growable-buffer INTERFACE builder-options)
+# Component: Layout Builder
+add_component(layout-builder)
+target_link_libraries(layout-builder INTERFACE growable-buffer builder-options)
+
+# Set target compile properties
+set_target_properties(
+  layout-builder
+  PROPERTIES CXX_STANDARD 17
+             CXX_STANDARD_REQUIRED YES
+             CXX_EXTENSIONS NO)
+set_target_properties(
+  builder-options growable-buffer
+  PROPERTIES CXX_STANDARD 11
+             CXX_STANDARD_REQUIRED YES
+             CXX_EXTENSIONS NO)
+
+# Build test suite?
+if(BUILD_TESTS)
+  add_subdirectory(tests)
+endif(BUILD_TESTS)
```

Comment:
I tries `cmake tests`:
```bash
(base) yana@iannas-macbook-pro-2 header-only % cmake tests
CMake Warning (dev) in CMakeLists.txt:
  No project() command is present.  The top-level CMakeLists.txt file must
  contain a literal, direct call to the project() command.  Add a line of
  code such as

    project(ProjectName)

  near the top of the file, but after cmake_minimum_required().

  CMake is pretending there is a "project(Project)" command on the first
  line.
This warning is for project developers.  Use -Wno-dev to suppress it.

-- The C compiler identification is AppleClang 13.0.0.13000029
-- The CXX compiler identification is AppleClang 13.0.0.13000029
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Warning (dev) in CMakeLists.txt:
  No cmake_minimum_required command is present.  A line of code such as

    cmake_minimum_required(VERSION 3.24)

  should be added at the top of the file.  The version specified may be lower
  if you wish to support older CMake versions for this project.  For more
  information run "cmake --help-policy CMP0000".
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Configuring done
CMake Warning (dev) at CMakeLists.txt:4 (add_executable):
  Policy CMP0063 is not set: Honor visibility properties for all target
  types.  Run "cmake --help-policy CMP0063" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  Target "test_1494-layout-builder" of type "EXECUTABLE" has the following
  visibility properties set for CXX:

    CXX_VISIBILITY_PRESET
    VISIBILITY_INLINES_HIDDEN

  For compatibility CMake is not honoring them for this target.
Call Stack (most recent call first):
  CMakeLists.txt:14 (addtest_nolibs)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at CMakeLists.txt:4 (add_executable):
  Policy CMP0063 is not set: Honor visibility properties for all target
  types.  Run "cmake --help-policy CMP0063" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  Target "test_1542-growable-buffer" of type "EXECUTABLE" has the following
  visibility properties set for CXX:

    CXX_VISIBILITY_PRESET
    VISIBILITY_INLINES_HIDDEN

  For compatibility CMake is not honoring them for this target.
Call Stack (most recent call first):
  CMakeLists.txt:15 (addtest_nolibs)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at CMakeLists.txt:4 (add_executable):
  Policy CMP0063 is not set: Honor visibility properties for all target
  types.  Run "cmake --help-policy CMP0063" for policy details.  Use the
  cmake_policy command to set the policy and suppress this warning.

  Target "test_1560-builder-options" of type "EXECUTABLE" has the following
  visibility properties set for CXX:

    CXX_VISIBILITY_PRESET
    VISIBILITY_INLINES_HIDDEN

  For compatibility CMake is not honoring them for this target.
Call Stack (most recent call first):
  CMakeLists.txt:16 (addtest_nolibs)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at CMakeLists.txt:18 (target_link_libraries):
  Policy CMP0028 is not set: Double colon in target name means ALIAS or
  IMPORTED target.  Run "cmake --help-policy CMP0028" for policy details.
  Use the cmake_policy command to set the policy and suppress this warning.

  Target "test_1494-layout-builder" links to:

    awkward::layout-builder

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at CMakeLists.txt:19 (target_link_libraries):
  Policy CMP0028 is not set: Double colon in target name means ALIAS or
  IMPORTED target.  Run "cmake --help-policy CMP0028" for policy details.
  Use the cmake_policy command to set the policy and suppress this warning.

  Target "test_1542-growable-buffer" links to:

    awkward::growable-buffer

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at CMakeLists.txt:20 (target_link_libraries):
  Policy CMP0028 is not set: Double colon in target name means ALIAS or
  IMPORTED target.  Run "cmake --help-policy CMP0028" for policy details.
  Use the cmake_policy command to set the policy and suppress this warning.

  Target "test_1560-builder-options" links to:

    awkward::builder-options

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

This warning is for project developers.  Use -Wno-dev to suppress it.

-- Generating done
-- Build files have been written to: /Users/yana/Projects/PR2280/awkward/header-only
(base) yana@iannas-macbook-pro-2 header-only % ls
CMakeCache.txt		CMakeFiles		CMakeLists.txt		Makefile		README.md		builder-options		cmake_install.cmake	growable-buffer		layout-builder		tests
(base) yana@iannas-macbook-pro-2 header-only % make
[ 16%] Building CXX object CMakeFiles/test_1494-layout-builder.dir/test_1494-layout-builder.o
/Users/yana/Projects/PR2280/awkward/header-only/tests/test_1494-layout-builder.cpp:3:10: fatal error: 'awkward/LayoutBuilder.h' file not found
#include "awkward/LayoutBuilder.h"
         ^~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
make[2]: *** [CMakeFiles/test_1494-layout-builder.dir/test_1494-layout-builder.o] Error 1
make[1]: *** [CMakeFiles/test_1494-layout-builder.dir/all] Error 2
make: *** [all] Error 2
(base) yana@iannas-macbook-pro-2 header-only % 

```

Reply:
>  I'm not sure I like the fine granularity for the headers, but it's just my personal view and I'm sure I'll get used to it :-) 

Hehe, I'm open to alternative approaches! I settled on this finer-grained solution so that we can specify CXX 14 and CXX 17 for the growable-buffer and layout-builder respectively. I can't (off the top of my head) think of any other way of doing this that let's us just re-use the build configuration between `awkward-cpp` and third party usage.

Reply:
Great! Thanks, it works for me.